src/backend/distributed/executor/multi_executor.c

```diff
@@ -24,273 +24,440 @@
+	EState *executorState = NULL;
+	MemoryContext executorTupleContext = NULL;
+	ExprContext *executorExpressionContext = NULL;
+	TupleDesc tupleDescriptor = NULL;
+	Relation fauxRelation = NULL;
+	ListCell *workerTaskCell = NULL;
+	uint32 columnCount = 0;
+	Datum *columnValues = NULL;
+	bool *columnNulls = NULL;
+	bool randomAccess = true;
+
+	executorState = scanState->customScanState.ss.ps.state;
+	executorTupleContext = GetPerTupleMemoryContext(executorState);
+	executorExpressionContext = GetPerTupleExprContext(executorState);
+
+	tupleDescriptor = customScanState.ss.ps.ps_ResultTupleSlot->tts_tupleDescriptor;
+	fauxRelation = FauxRelation(tupleDescriptor);
+
+	columnCount = tupleDescriptor->natts;
+	columnValues = palloc0(columnCount * sizeof(Datum));
+	columnNulls = palloc0(columnCount * sizeof(bool));
+
+	Assert(scanState->tuplestorestate == NULL);
+	scanState->tuplestorestate = tuplestore_begin_heap(randomAccess, false, work_mem);
+
+	foreach(workerTaskCell, workerTaskList)
 	{
-		/* drop into the standard executor */
-		standard_ExecutorFinish(queryDesc);
+		Task *workerTask = (Task *) lfirst(workerTaskCell);
+		StringInfo jobDirectoryName = NULL;
+		StringInfo taskFilename = NULL;
+		List *copyOptions = NIL;
+		CopyState copyState = NULL;
+
+		jobDirectoryName = MasterJobDirectoryName(workerTask->jobId);
+		taskFilename = TaskFilename(jobDirectoryName, workerTask->taskId);
+
+		if (BinaryMasterCopyFormat)
+		{
+			DefElem *copyOption = makeDefElem("format", (Node *) makeString("binary"));
+			copyOptions = lappend(copyOptions, copyOption);
+		}
+
+		copyState = BeginCopyFrom(fauxRelation, taskFilename->data, false, NULL,
+								  copyOptions);
+
+		while (true)
+		{
+			MemoryContext oldContext = NULL;
+			bool nextRowFound = false;
+
+			ResetPerTupleExprContext(executorState);
+			oldContext = MemoryContextSwitchTo(executorTupleContext);
+
+			nextRowFound = NextCopyFrom(copyState, executorExpressionContext,
+										columnValues, columnNulls, NULL);
+			if (!nextRowFound)
+			{
+				MemoryContextSwitchTo(oldContext);
+				break;
+			}
+
+			tuplestore_putvalues(scanState->tuplestorestate, tupleDescriptor,
+								 columnValues, columnNulls);
+			MemoryContextSwitchTo(oldContext);
+		}
+
+		EndCopyFrom(copyState);
 	}
 }
 
 
 /*
- * multi_ExecutorEnd is a hook called to deallocate resources used during
- * query execution.
- *
- * If the query executed was the portion of a distributed query running on the
- * master, remove the resources that were needed for distributed execution.
+ * FauxRelation creates a faux Relation from the given tuple descriptor.
+ * To be able to use copy.c, we need a Relation descriptor. As there is no
+ * relation corresponding to the data loaded from workers, we need to fake one.
+ * We just need the bare minimal set of fields accessed by BeginCopyFrom().
  */
-void
-multi_ExecutorEnd(QueryDesc *queryDesc)
+static Relation
+FauxRelation(TupleDesc tupleDescriptor)
```

Comment:
_Faux_'s a bit of a weird modifier here. Can we go with `Stub` (more common in programming) or `Fake` (doesn't mix linguistic roots)?